manager/assets/modext/widgets/core/modx.grid.js

```diff
@@ -495,14 +515,190 @@ Ext.extend(MODx.grid.Grid,Ext.grid.EditorGridPanel,{
+     * rely on a pair of permissions (save_x and new_x), both of which must be enabled to create/duplicate a grid item
+     *
+     * @return void
+     */
+    ,setUserCanCreate: function(permissions) {
+        permissions = permissions.map(item => item.trim());
+        this.userCanCreate = permissions.every(permission => MODx.perm[permission]);
+    }
+
+    /**
+     * @property {Function} setUserCanDelete - Assigns a value to userCanDelete property based on
+     * the user's permissions; used to adjust which menu items are available in the context menus
+     * and whether to render the Delete menu item within a grid toolbar's Batch button
+     *
+     * @param {Array} permissions - A set of permissions keys to evaluate
+     *
+     * @return void
+     */
+    ,setUserCanDelete: function(permissions) {
+        permissions = permissions.map(item => item.trim());
+        this.userCanDelete = permissions.every(permission => MODx.perm[permission]);
+    }
+
+
+    /* Record-Level Permissions Checks, for objects with specific policies */
+
+    ,userCanEditRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && permissions.update === true ? true : false ;
+    }
+
+    ,userCanDeleteRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && !record.json.isProtected && permissions.delete === true ? true : false ;
+    }
+
+    ,userCanDuplicateRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && permissions.duplicate === true ? true : false ;
+    }
+
+    /**
+     * @property {Function} setShowActionsMenu - Based on properties set in the calling child class and the
+     * the current user's permissions for actions taken within that class (create, edit, delete, etc),
+     * evaluates whether the actions menu trigger should appear and sets boolean value on the showActionsMenu property
+     *
+     * @return void
+     */
+    ,setShowActionsMenu: function() {
+        if (this.config.disableContextMenuAction === true) {
+            this.showActionsMenu = false;
+            return;
+        }
+        const permissions = [];
+        this.gridMenuActions.forEach(mode => {
+            mode = mode == 'duplicate' ? 'userCanCreate' : 'userCan' + Ext.util.Format.capitalize(mode);
+            const modePermission = mode === 'userCanExport' ? true : this[mode] ;
+            if (['userCanCreate', 'userCanEdit'].includes(mode) && modePermission === true) {
+                this.hasSavePermissions = true;
+            }
+            permissions.push(modePermission);
+        });
+        this.showActionsMenu = permissions.length === 0 || permissions.every(permission => permission === false) === true
+            ? false
+            : true
+            ;
```

Comment:
Also it would be useful to cover such code with unit-tests.
```suggestion
        this.showActionsMenu = !(!permissions.length || permissions.every(permission => !permission)) ;
```

Reply:
I'm going to leave this one as-is, as I find it more readable.